freedomofpress/qubes-template-securedrop-workstation
# Initial support for Debian Bullseye templates

## Overview
Towards https://github.com/freedomofpress/securedrop-workstation/issues/733

The work here constitutes a spike for Bullseye (and 4.1) compatibility. As usual, there were some minor touchups required to the build script, to ensure compatibility with the upstream `qubes-builder` logic (https://github.com/QubesOS/qubes-builder). It's important to note this is just a "spike" because full support will require additional changes, noted below.


## Not implemented
  * rebuilding all component packages for Bullseye (specifically for python3.9, rather than python3.7), and hosting those packages in the FPF apt repo.
  * adding a prod-signed tag to this repo (thus the scary "TEMPORARY: disables signing verification" commit)
  * rehosting paxctld package in FPF apt repo (it was pulled from Debian Stable; see https://github.com/freedomofpress/securedrop-workstation/issues/733#issuecomment-1084115175 for discussion about options)
 
## Test plan

Confirm that you can build a Bullseye-based template RPM using the steps below. Since there are several changes in other repos, you'll need to build a few packages elsewhere, first.


1. Check out this kernel builder PR: https://github.com/freedomofpress/kernel-builder/pull/20
2. Run `make securedrop-workstation 5.15`, and go grab a coffee. When the build finishes, store the tarball for safe keeping. Copy the `linux-image` and `linux-headers` packages to your `sd-template-builder` VM.
3. Checkout this packaging PR: https://github.com/freedomofpress/securedrop-debian-packaging/pull/301
4. Update the metapackage version to match the version of the kernel that you built above (it'll be newer than 5.15.33, documented here). 
5. Run `PKG_VERSION=5.15.<YOUR_VERSION> PKG_PLATFORM=bullseye make securedrop-workstation-grsec`. Copy that metapackage to the `sd-template-builder` VM.
6. Fetch the latest version of paxctld from the grsecurity website: https://grsecurity.net/download/ . Copy that package to `sd-template-builder`.
7. Check out https://github.com/freedomofpress/qubes-template-securedrop-workstation/pull/24 locally. Edit the package versions and paths in `
securedrop-workstation/04_install_qubes_post.sh` to match what you have locally.
8. Run `make template`, grab another coffee.
- [ ] :red_square: **Ensure that signatures are verified!** That is, remove the commit "TEMPORARY: disable signing verification" (https://github.com/freedomofpress/qubes-template-securedrop-workstation/pull/24/commits/43ecc16c4cb01a98ffcaf5a5666fd15b17b0f786).

After the build completes, you'll have a fresh RPM named something like: 
`qubes-template-securedrop-workstation-bullseye-4.0.6-202204132331.noarch.rpm`.

1. Copy that RPM to dom0: `qvm-run -p sd-kernel-builder path/to/qubes-template-securedrop-workstation-bullseye-4.0.6-202204132331.noarch.rpm` > qubes-template-securedrop-workstation-bullseye-4.0.6-202204132331.noarch.rpm
2. Install it: `sudo dnf install ./qubes-template-securedrop-workstation-bullseye-4.0.6-202204132331.noarch.rpm`
3. Modify prefs: `qvm-prefs securedrop-template-bullseye virt_mode hvm && qvm-prefs securedrop-template-bullseye kernel ''`.
4. Boot VM. Confirm via `uname -a` that it's using the expected custom kernel version. Run `apt-get update and `apt-get upgrade`. 

## Next steps
There aren't any SDW deb packages ready for Bullseye yet, so the virtualenvs for packages like `securedrop-client` won't work, even if you install them. More work required on that front.

For the Template VM, once we're happy with the logic, we'll need to **remove** the temporary commits that disable signing verification on this PR, and push a new prod-signed tag. (N.B. great opportunity to use a _keyring_ of SD maintainers, rather than the Release Key!) Then `make template` will work after verifying the integrity of all the resources it pulls in, including this repo.
